uniffi/src/ffi/rustcalls.rs

```diff
@@ -0,0 +1,218 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! # Low-level support for calling rust functions
+//!
+//! This module helps the scaffolding code make calls to rust functions and pass back the result to the FFI bindings code.
+//!
+//! It handles:
+//!    - Catching panics
+//!    - Adapting `Result<>` types into either a return value or an error
+
+use crate::{RustBuffer, ViaFfi};
+use anyhow::Result;
+use ffi_support::IntoFfi;
+use std::panic;
+use std::mem::MaybeUninit;
+
+/// Represents the success/error of a rust call
+///
+/// ## Usage
+///
+/// - The consumer code allocates a `RustCallStatus`
+///    - The error_buf field can either be uninitialized or initialized to zero-capacity and a null
+///      data buffer.  The data here is going to be overwritten so any allocations will leak memory.
```

Comment:
So basically the way we're using `RustBuffer` is:
  - The consumer passes in a zeroed out rust buffer struct
  - In the case of an Error, we overwrite that data with a `RustBuffer` that contains the error.

Is this usage supported or should it be?  If so I think we should update the `RustBuffer` documentation to mention that.  If not, is there a better way to achieve this?

Reply:
That makes a lot of sense.  For some reason I was thinking that `RustBuffer`s were intended to be immutable, but after re-reading the docs I realize that's not the case.  Is it generally true that passing in a `RustBuffer` to a function means transferring ownership and therefore the calling code shouldn't try to access it once they pass it in?  If so, maybe I'll add a bullet point to the `RustBuffer` doc comment about that.